dao/tenant_dao.go

```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
 
 	return &tenant, nil
 }
+
+func (t *tenantDaoImpl) GetUntranslatedTenants() ([]m.Tenant, error) {
+	var tenants []m.Tenant
+
+	err := DB.Debug().
+		Model(&m.Tenant{}).
+		Where(`("external_tenant" IS NOT NULL OR "external_tenant" != '') AND ("org_id" IS NULL OR "org_id" = '')`).
+		Find(&tenants).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tenants, err
+}
+
+func (t *tenantDaoImpl) TranslateTenants() (int64, uint64, uint64, []m.TenantTranslation, error) {
+	// The where condition to fetch only the untranslated tenants.
+	untranslatedTenantsWhereCondition := `("external_tenant" IS NOT NULL OR "external_tenant" != '') AND ("org_id" IS NULL OR "org_id" = '')`
+
+	// Count the total number of translatable tenants.
+	var translatableTenants int64
+	// Count the translation operations.
+	var translatedTenants, untranslatedTenants uint64
+	// Keep track of each of the tenant translations.
+	var tenantTranslations []m.TenantTranslation
+
+	// Run everything inside a transaction to avoid inconsistencies. This way if new tenants are added/removed while
+	// the translation is in progress, all the counts and the results will still match.
+	err := DB.Debug().Transaction(func(tx *gorm.DB) error {
+		// Get the count to be able to process the tenants in batches of 100 tenants.
+		err := tx.
+			Model(&m.Tenant{}).
+			Where(untranslatedTenantsWhereCondition).
+			Count(&translatableTenants).
+			Error
+
+		if err != nil {
+			return fmt.Errorf("unable to count untranslated tenants: %w", err)
+		}
+
+		translator := tenantid.NewTranslator(config.Get().TenantTranslatorUrl)
+		for i := int64(0); i < translatableTenants; i += 100 {
+			// Grab all the EBS account numbers to be processed.
+			var ebsAccountNumbers []string
+			err := tx.
+				Model(&m.Tenant{}).
+				Where(untranslatedTenantsWhereCondition).
+				Pluck("external_tenant", &ebsAccountNumbers).
+				Offset(int(i)).
+				Limit(100).
+				Error
+
+			if err != nil {
+				return fmt.Errorf("unable to fetch the tenants to translate them: %w", err)
+			}
+
+			// Batch process 100 EBS account numbers.
+			results, err := translator.EANsToOrgIDs(context.Background(), ebsAccountNumbers)
+			if err != nil {
+				return fmt.Errorf("error when calling the translation service: %w", err)
+			}
+
+			for _, result := range results {
+				var tenantTranslation = m.TenantTranslation{}
+
+				// Did the translation service return an error?
+				if result.Err != nil {
+					logging.Log.WithField("external_tenant", *result.EAN).Errorf(`could not translate to "org_id": %s`, err)
+
+					tenantTranslation.ExternalTenant = *result.EAN
+					tenantTranslation.Error = err
+					tenantTranslations = append(tenantTranslations, tenantTranslation)
+
+					untranslatedTenants++
+					continue
+				}
+
+				// Grab the updated tenant for the "translated tenant" struct.
+				var tenant m.Tenant
+				// Update the tenant with the translated "orgId".
+				dbResult := tx.
```

Comment:
Is in postgres batch update ?

Reply:
> Who will call this ? Will it be in cronjob ?

More like CronJacob, since he'll be the one sending these requests for the translations

Reply:
I thought Gorm supported batch updating, but I am facing a really weird "unsupported data" when fetching the tenants and then calling `db.Updates(&tenants)`. I've looked on the internet to actually see if you can do a batch update when you're trying to update multiple rows with different values and... things are a bit complicated.

However, aside from this, the implementation I came up with was a little bit weird:

1. I fetched 100 tenants per loop.
2. Created a `map[string]*m.Tenant` to store in memory the tenants by their external tenant.
3. Called the translation service.
4. I used the `map[string]` to update each of the tenants with their corresponding `org_id`.
5. Updated the database with the tenants.
6. Looped through the tenants to generate a proper HTTP response.

Here's the implementation: https://gist.github.com/MikelAlejoBR/54c6f3d3694556cc4344f6f8c3e03f88

I would honestly not refactor the current PR's code to the above, to be honest... plus I would have to check if there's a way to properly batch update with Gorm without hitting an error :eyes:

Reply:
yes, you are right and thank you very much for trying it and for those information! 👏